### PR TITLE
Adding commands+shortcuts for keyboard nav of tabs in dock panel.

### LINKED ISSUE #1637: Feature request: Keyboard Shortcut to navigate to next tab / previous tab
Most operating system GUIs, browsers, editors, etc. have a method for navigating using the keyboard only, without the need to use the mouse.  With JupyterLab, I find myself frequently switching between tabs using my mouse, but am wishing there was a keyboard shortcut to navigate to the next or previous tab.

If this capability were to be added to JupyterLab, a Keyboard Shortcut for each of the two actions would need to be chosen.  This Keyboard Shortcut should not conflict with any other Keyboard Shortcut that may already be getting caught and triggering some action as a result.

Here's an example of the navigation action I am describing.  On most browsers, editors, etc. the Keyboard shortcuts Ctrl + PgUp and Ctrl + PgDn perform the Next Tab and Previous Tab actions.   So this Keyboard shortcut is already reserved, but indicates PgUp / PgDn are part of the Keyboard shortcut for navigating tabs.

I don't think Alt + PgUp and Alt + PgDn are already used by the browser or operating system, ... so perhaps that would be a good Keystroke combination that works for this.    (The Chrome browser, and possibly others) do trap [Command] + PgUp and [Command] + PgDn for scrolling within a text area, as an alternative to the mouse wheel, so using the [Command] key (also known as the "Open Apple", "Window", or "Super" key, depending on operating system) would likely be a bad choice.

I did see Google Mail (gmail) has an option in Settings to enable or disable Keyboard shortcuts (with the default being "Keyboard Shortcuts Off").   https://support.google.com/mail/answer/6594?hl=en  I'm not sure such a setting would be necessary for JupyterLab, but wanted to share how gmail handles the subject.   Incidentally the Keyboard shortcut for navigating their "Section" (tabs) is the "tick" ( ` ) and tilde ( ~ ) for the Next Section / Previous Section actions.

A Wikipedia article on common shortcuts is here: 
https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts#Tab_management

The documentation for a popular editor's shortcuts is here: 
http://docs.sublimetext.info/en/latest/reference/keyboard_shortcuts_win.html

There is one wrinkle to this introduced by Jupyterlab though.  Attached is a screen shot [provided](https://gitter.im/jupyterlab/jupyterlab?at=58985c5d1465c46a562a8747) by Afshin Darian showing how the main panel area can have more than one sub panel section.  So the Next tab / Previous Tab might only apply to the current paenl area.  Or the Next Tab action when already at the rightmost-tab of the current panel area might just go to the first Tab in next panel area.     I wasn't able to find another app with similar panel areas to be able to make an educated guess as to which approach might seem better, and am not a UX professional anyway. 

Tl;dr: Add some keyboard shortcut to handle Next tab / Previous tab within JupyterLab.

screen-shot-2017-02-06-at-11 21 35-am

### PR DESCRIPTION
**Pull request by ellisonbg**
Fixes #1637 

This works, but I have a few questions:

* Have got the logic right for these methods on `ApplicationShell`?
* What keyboard shortcuts to use (currently accel left+right)?
* Right now the terminal swallows these keyboard shortcuts, how do I make the terminal ignore them.

### CODE REVIEW

**Comment by ellisonbg** [comment]
@sccolbert for review

Another question: right now I don't do anything if a user tries to select the next tab and they are at the end of the active panel. Is there a reasonable way of iterating through the panels in the dock panel?

**Comment by ellisonbg** [comment]
I think I see the logic for going between bars in the `dockPanel.tabBars()` to switch between panels.   Another question though.

I have the `TabBar` instance that contains the active widget/tab. I want to find the next or previous `TabBar` from the `dockPanel.tabBars()` iterator. What is the best way of doing that with arrays or sequences in phosphor?

**Comment by blink1073** [comment]
We usually call [toArray](https://github.com/phosphorjs/phosphor/blob/47fce2da39ad2397af17b0e01a35c7560575939a/src/algorithm/iteration.ts#L133) if were are going to do any indexing on an iterator.  
See https://github.com/phosphorjs/phosphor/issues/185 for how we are going to handle capturing keydown events before input areas get them.

**Comment by ellisonbg** [comment]
Woohoo!! I have fun with this one. It is all working. This makes it really easy to navigate the entire dock panel using `Accel ArrowLeft` and `Accel ArrowRight`. It will cycle through tabs in each panel and also through panels, including periodic boundary conditions. Should be ready to go!

**Comment by blink1073** [comment]
Squeaked it in for 0.16!  I'll test/review this now.

File: src/application/shell.ts [hunk 1]
```diff
@@ -341,6 +386,60 @@ class ApplicationShell extends Widget {
...
+    return void 0;
+  }
+
+  /*
+   * Return the TabBar previous to the current TabBar (see above) or undefined.
+   */
+  private _previousTabBar(): TabBar {
+    let current = this._currentTabBar();
+    if (current) {
+      let bars = toArray<TabBar>(this._dockPanel.tabBars());
```
Comment (sccolbert) [inline src/application/shell.ts]: the `<TabBar>` type is not needed. `toArray` will automatically infer the generic type based on the argument.

File: src/application/shell.ts [hunk 2]
```diff
@@ -341,6 +386,60 @@ class ApplicationShell extends Widget {
...
+    return void 0;
+  }
+
+  /*
+   * Return the TabBar next to the current TabBar (see above) or undefined.
+   */
+  private _nextTabBar(): TabBar {
+    let current = this._currentTabBar();
+    if (current) {
+      let bars = toArray<TabBar>(this._dockPanel.tabBars());
```
Comment (sccolbert) [inline src/application/shell.ts]: ditto

File: src/shortcuts/plugin.ts [hunk 3]
```diff
@@ -72,6 +72,16 @@ const SHORTCUTS = [
...
   },
   {
+    command: ApplicationCommandIDs.activateNextTab,
+    selector: 'body',
+    keys: ['Accel ArrowRight']
+  },
+  {
+    command: ApplicationCommandIDs.activatePreviousTab,
+    selector: 'body',
+    keys: ['Accel ArrowLeft']
```
Comment (sccolbert) [inline src/shortcuts/plugin.ts]: do these shortcuts not conflict with anything else in the app?

**Comment by sccolbert** [comment]
Made a few comments. Looks reasonable.

File: src/application/shell.ts [hunk 4]
```diff
@@ -308,6 +308,51 @@ class ApplicationShell extends Widget {
...
 
+  /*
+   * Activate the next Tab in the active TabBar.
+  */
+  activateNextTab(): void {
+    let current = this._currentTabBar();
+    if (current) {
+      let ci = current.currentIndex;
+      if (ci !== -1) {
+        if (ci < (current.titles.length-1)) {
```
Comment (blink1073) [inline src/application/shell.ts]: spaces around the binary operator

File: src/application/shell.ts [hunk 5]
```diff
@@ -308,6 +308,51 @@ class ApplicationShell extends Widget {
...
+          current.currentIndex += 1;
+          current.currentTitle.owner.activate();
+        } else if (ci === (current.titles.length-1)) {
```
Comment (blink1073) [inline src/application/shell.ts]: idem

File: src/application/shell.ts [hunk 6]
```diff
@@ -308,6 +308,51 @@ class ApplicationShell extends Widget {
...
+      let ci = current.currentIndex;
+      if (ci !== -1) {
+        if (ci > 0) {
+          current.currentIndex -= 1;
+          current.currentTitle.owner.activate();
+        } else if (ci === 0) {
+          let prevBar = this._previousTabBar();
+          if (prevBar) {
+            let len = prevBar.titles.length;
+            prevBar.currentIndex = len-1;
```
Comment (blink1073) [inline src/application/shell.ts]: idem

File: src/application/shell.ts [hunk 7]
```diff
@@ -341,6 +386,60 @@ class ApplicationShell extends Widget {
...
+   */
+  private _previousTabBar(): TabBar {
+    let current = this._currentTabBar();
+    if (current) {
+      let bars = toArray<TabBar>(this._dockPanel.tabBars());
+      let len = bars.length;
+      let ci = bars.indexOf(current);
+      let prevBar: TabBar = null;
+      if (ci > 0) {
+        prevBar = bars[ci-1];
```
Comment (blink1073) [inline src/application/shell.ts]: idem (and two lines down)

File: src/application/shell.ts [hunk 8]
```diff
@@ -341,6 +386,60 @@ class ApplicationShell extends Widget {
...
+   * Return the TabBar next to the current TabBar (see above) or undefined.
+   */
+  private _nextTabBar(): TabBar {
+    let current = this._currentTabBar();
+    if (current) {
+      let bars = toArray<TabBar>(this._dockPanel.tabBars());
+      let len = bars.length;
+      let ci = bars.indexOf(current);
+      let nextBar: TabBar = null;
+      if (ci < (len-1)) {
```
Comment (blink1073) [inline src/application/shell.ts]: more idems

**Comment by blink1073** [comment]
This is *really* elegant, well done!

File: src/application/shell.ts (hunk 4 above)
Comment (sccolbert) [inline src/application/shell.ts]: While we're being nit-picky, you don't need the extra parens either. The `-` operator has a higher precedence than logical operators: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence

File: src/shortcuts/plugin.ts (hunk 3 above)
Comment (ellisonbg) [inline src/shortcuts/plugin.ts]: Not that I know of...

Comment (blink1073) [inline src/shortcuts/plugin.ts]: Seconded.

File: src/application/shell.ts (hunk 4 above)
Comment (ellisonbg) [inline src/application/shell.ts]: I will remove the extra parens. Can you clarify what you mean about the spaces around the binary operator?

Comment (blink1073) [inline src/application/shell.ts]: `current.titles.length-1` -> `current.titles.length - 1`

Comment (ellisonbg) [inline src/application/shell.ts]: Nevermind, I understand...

**Comment by ellisonbg** [comment]
K, should be good, thanks for the review!

**Comment by blink1073** [comment]
🎉
